functionary/train/train_lora.py

```diff
@@ -0,0 +1,396 @@
+import json
+import math
+import pathlib
+import random
+from dataclasses import dataclass, field
+from typing import Dict, Optional
+
+import torch
+import torch.distributed
+from torch.nn import CrossEntropyLoss
+import sys
+import os
+
+os.environ["WANDB_LOG_MODEL"] = "all"
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "../.."))
+from functionary.prompt import EndToken
+from functionary.train.custom_datasets import CustomDataset
+
+from deepspeed import zero
+from deepspeed.runtime.zero.partition_parameters import ZeroParamStatus
+
+import transformers
+from transformers import LlamaTokenizer, Trainer, BitsAndBytesConfig, deepspeed
+from transformers import LlamaForCausalLM
+from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training
+import bitsandbytes as bnb
+
+LOCAL_RANK = int(os.getenv("LOCAL_RANK", "0"))
+
+
+def print_rank0(*arg):
+    if LOCAL_RANK == 0:
+        print(*arg)
+
+
+@dataclass
+class ModelArguments:
+    model_name_or_path: Optional[str] = field(default="meta-llama/Llama-2-7b-hf")
+
+
+@dataclass
+class DataArguments:
+    train_data_path: str = field(default=None, metadata={"help": "Path to the training data."})
+    eval_data_path: str = field(default=None, metadata={"help": "Path to the eval data."})
+
+
+@dataclass
+class TrainingArguments(transformers.TrainingArguments):
+    cache_dir: Optional[str] = field(default=None)
+    model_max_length: int = field(
+        default=4096,
+        metadata={"help": "Maximum sequence length. Sequences will be right padded (and possibly truncated)."},
+    )
+
+
+def safe_save_model_for_hf_trainer(trainer: transformers.Trainer, output_dir: str):
```

Comment:
Let's use the new saving API that we used in other train.py

Reply:
Actually I wrote the code from train.py  so include this function, but we don't need to use it, so I will delete